File: npm/vite-plugin-musea/src/preview/html.ts
/**
 * Preview HTML page generation for Musea component previews.
 *
 * Generates the HTML wrapper pages that host Vue component previews
 * inside iframes.
 *
 * Extracted from preview.ts to keep file sizes manageable.
 */

import type { ArtFileInfo, ArtVariant } from "../types/index.js";
import { escapeHtml } from "../utils.js";

export function generatePreviewHtml(
  art: ArtFileInfo,
  variant: ArtVariant,
  _basePath: string,
  viteBase?: string,
): string {
  // Use preview-module HTTP endpoint instead of virtual module import.
  // Virtual module imports in inline scripts require transformIndexHtml,
  // which creates malformed html-proxy URLs when the page URL has query params.
  const previewModuleUrl = `${_basePath}/preview-module?art=${encodeURIComponent(art.path)}&variant=${encodeURIComponent(variant.name)}`;
  const base = (viteBase || "/").replace(/\/$/, "");

  return `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>${escapeHtml(art.metadata.title)} - ${escapeHtml(variant.name)}</title>
  <script type="module" src="${base}/@vite/client"></script>
  <style>
    * { box-sizing: border-box; margin: 0; padding: 0; }
    :root {
      --musea-preview-padding: clamp(1rem, 2vw, 1.5rem);
    }
    html, body {
      width: 100%;
      height: 100%;
    }
    body {
      font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif;
      background: #ffffff;
      padding: var(--musea-preview-padding);
      overflow: auto;
    }
    .musea-variant {
      width: 100%;
      min-height: calc(100vh - (var(--musea-preview-padding) * 2));
    }
    .musea-error {
      color: #dc2626;
      background: #fef2f2;
      border: 1px solid #fecaca;
      border-radius: 8px;
      padding: 1rem;
      font-size: 0.875rem;
      max-width: 400px;
    }
    .musea-error-title {
      font-weight: 600;
      margin-bottom: 0.5rem;
    }
    .musea-error pre {
      font-family: monospace;
      font-size: 0.75rem;
      white-space: pre-wrap;
      word-break: break-all;
      margin-top: 0.5rem;
      padding: 0.5rem;
      background: #fff;
      border-radius: 4px;
    }
    .musea-loading {
      display: flex;
      align-items: center;
      gap: 0.75rem;
      color: #6b7280;
      font-size: 0.875rem;
    }
    .musea-spinner {
      width: 20px;
      height: 20px;
      border: 2px solid #e5e7eb;
      border-top-color: #3b82f6;
      border-radius: 50%;
      animation: spin 0.8s linear infinite;
    }
    @keyframes spin { to { transform: rotate(360deg); } }

    /* Musea Addons: Checkerboard background for transparent mode */
    .musea-bg-checkerboard {
      background-image:
        linear-gradient(45deg, #ccc 25%, transparent 25%),
        linear-gradient(-45deg, #ccc 25%, transparent 25%),
        linear-gradient(45deg, transparent 75%, #ccc 75%),
        linear-gradient(-45deg, transparent 75%, #ccc 75%) !important;
      background-size: 20px 20px !important;
      background-position: 0 0, 0 10px, 10px -10px, -10px 0 !important;
    }

    /* Musea Addons: Measure label */
    .musea-measure-label {
      position: fixed;
      background: #333;
      color: #fff;
      font-size: 11px;
      padding: 2px 6px;
      border-radius: 3px;
      pointer-events: none;
      z-index: 100000;
    }
  </style>
</head>
<body>
  <div id="app" class="musea-variant" data-art="${escapeHtml(art.path)}" data-variant="${escapeHtml(variant.name)}">
    <div class="musea-loading">
      <div class="musea-spinner"></div>
      Loading component...
    </div>
  </div>
  <script type="module" src="${previewModuleUrl}"></script>
</body>
</html>`;
}

File: crates/vize_patina/src/output/html.rs
//! Self-contained HTML report.

use crate::linter::LintResult;
use crate::output::format_summary;
use crate::output::shared::{diagnostic_views, escape_html, result_counts};
use vize_carton::{String, append};

pub(super) fn format_html(results: &[LintResult], sources: &[(String, String)]) -> String {
    let views = diagnostic_views(results, sources);
    let (errors, warnings) = result_counts(results);
    let mut output = String::from(
        "<!doctype html>\n<html lang=\"en\">\n<head>\n<meta charset=\"utf-8\">\n<title>Patina Lint Report</title>\n<style>\nbody{font-family:system-ui,-apple-system,BlinkMacSystemFont,\"Segoe UI\",sans-serif;margin:2rem;line-height:1.5;color:#191919;background:#fff}main{max-width:960px}.summary{color:#555}.file{margin-top:2rem}.diagnostic{border:1px solid #ddd;border-left-width:4px;border-radius:6px;padding:1rem;margin:1rem 0}.diagnostic.error{border-left-color:#c62828}.diagnostic.warning{border-left-color:#ad6b00}.meta{display:flex;gap:.75rem;flex-wrap:wrap;align-items:center}.severity{text-transform:uppercase;font-size:.78rem;font-weight:700}.location,.docs{color:#666}code,pre{font-family:ui-monospace,SFMono-Regular,Menlo,Consolas,monospace}pre{white-space:pre-wrap;background:#f7f7f7;padding:.75rem;border-radius:4px;overflow:auto}\n</style>\n</head>\n<body>\n<main>\n<h1>Patina Lint Report</h1>\n",
    );
    output.push_str("<p class=\"summary\">");
    output.push_str(&escape_html(&format_summary(
        errors,
        warnings,
        results.len(),
    )));
    output.push_str("</p>\n");

    if views.is_empty() {
        output.push_str("</main>\n</body>\n</html>\n");
        return output;
    }

    let mut current_file = "";
    for view in &views {
        if current_file != view.file {
            if !current_file.is_empty() {
                output.push_str("</section>\n");
            }
            current_file = view.file;
            output.push_str("<section class=\"file\">\n<h2>");
            output.push_str(&escape_html(current_file));
            output.push_str("</h2>\n");
        }

        output.push_str("<article class=\"diagnostic ");
        output.push_str(view.severity_name());
        output.push_str("\">\n<header class=\"meta\"><span class=\"severity\">");
        output.push_str(view.severity_name());
        output.push_str("</span><code>");
        output.push_str(&escape_html(view.rule_id));
        output.push_str("</code><span class=\"location\">");
        append!(
            output,
            "{}:{}-{}:{}",
            view.line,
            view.column,
            view.end_line,
            view.end_column
        );
        output.push_str("</span></header>\n<p>");
        output.push_str(&escape_html(view.message));
        output.push_str("</p>\n<p class=\"docs\">Reference: <code>");
        output.push_str(&escape_html(view.rule_docs_path));
        output.push_str("</code></p>\n");

        if let Some(help) = view.help_text.as_deref() {
            output.push_str("<pre>");
            output.push_str(&escape_html(help));
            output.push_str("</pre>\n");
        }

        output.push_str("</article>\n");
    }

    output.push_str("</section>\n</main>\n</body>\n</html>\n");
    output
}
